components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java

```diff
@@ -522,8 +522,16 @@ public String getNormalizedRequestURI(String requestURI) throws URISyntaxExcepti
             return null;
         }
 
-        String decodedURl = URLDecoder.decode(requestURI, StandardCharsets.UTF_8.name());
-        return new URI(decodedURl).normalize().toString();
+        // Decode once
+        String decodedUri = URLDecoder.decode(requestURI, StandardCharsets.UTF_8.name());
+
+        // If the decoded URI still contains '%', consider it unsafe
+        if (decodedUri.contains("%")) {
```

Comment:
Shall we use regex `.*%[0-9a-fA-F]{2}.*` here?

This specifically looks for a percent sign followed by two hexadecimal digits. A valid percent-encoded character always follows the pattern %XX where XX is hex (00–FF).

If we only check %, a valid query parameter value also might get blocked.